src/core_editor/edit_stack.rs

```diff
@@ -0,0 +1,144 @@
+/// This outlines the operations (interface) that any data strcuture will need to obey to be useful
+/// to the core_editor as undo/redo stack
+pub trait EditStack<T>: Send
+where
+    T: Default,
+    T: Clone,
+    T: Send,
+{
+    /// Go back one point in the undo stack.
+    /// If present on first edit do nothing
+    fn undo(&mut self) -> &T;
+
+    /// Go forward one point in the undo stack.
+    /// If present on the last edit do nothing
+    fn redo(&mut self) -> &T;
+
+    /// Insert a new entry to the undo stack.
+    /// NOTE: (IMP): If we have hit undo a few times then discard all the other values that come
+    /// after the current point
+    fn insert(&mut self, value: T);
+
+    /// Reset the stack to the initial state
+    fn reset(&mut self);
+
+    /// Return the entry currently being pointed to
+    fn current(&mut self) -> &T;
+
+    /// List out all the entries on the undo stack
+    /// Mostly used for debugging. Might remove this one
+    fn edits<'a>(&'a self) -> Box<dyn Iterator<Item = &'a T> + 'a>;
+}
```

Comment:
This should not be a trait. There is no requirement for multiple coexisting impls and it makes encapsulation by visibility harder. Concrete types are also easier when testing alternative implementations as testing and reviewing a patch without changes to the unit tests is nicer (Instead of having to copy and slightly change the test suite and being unsure if it is testing the same thing. Dang that traits are not concepts in this case :P)